{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f5172de8",
   "metadata": {},
   "source": [
    "# Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94334433",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.302755Z",
     "start_time": "2023-09-05T18:24:27.924781Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Data manipulation\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Data science\n",
    "import math\n",
    "import scipy.stats as stats\n",
    "from sklearn.experimental import enable_iterative_imputer\n",
    "from sklearn.impute import IterativeImputer\n",
    "from statsmodels.stats.multitest import multipletests as mt\n",
    "\n",
    "# Plots\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "\n",
    "# Working with dates\n",
    "from datetime import date,datetime\n",
    "import dateutil\n",
    "\n",
    "# Looping  progress\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "# Reg expressions\n",
    "import re\n",
    "\n",
    "# Pretty table printing\n",
    "import tabulate\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "# Misc libraries\n",
    "from IPython.display import display, HTML\n",
    "#from IPython.core.display import display, HTML\n",
    "\n",
    "# Set seaborn figure size, font size, and style\n",
    "sns.set(rc={'figure.figsize':(11.7,8.27)})\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")\n",
    "\n",
    "# Set Pandas options so we can see our entire dataframe\n",
    "pd.options.display.max_rows = 10000\n",
    "pd.options.display.max_columns = 10000\n",
    "pd.options.display.max_colwidth = None\n",
    "\n",
    "# Print our versions of this packages, this allows us to make sure\n",
    "# we have the working versions we need. \n",
    "print(f\"Pandas version: {pd.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecba0d9f",
   "metadata": {},
   "source": [
    "# Prep Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8243cea0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.316936Z",
     "start_time": "2023-09-05T18:24:30.304753Z"
    }
   },
   "outputs": [],
   "source": [
    "# Remove grey side bars\n",
    "display(HTML(\"<style>.container { width:90% !important; }</style>\"))\n",
    "\n",
    "os.chdir('../../../../results/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9104e9b4",
   "metadata": {},
   "source": [
    "# Make 4 panel plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fd2ebf",
   "metadata": {},
   "source": [
    "## Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a471b67d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.342940Z",
     "start_time": "2023-09-05T18:24:30.318941Z"
    },
    "code_folding": [
     37
    ]
   },
   "outputs": [],
   "source": [
    "# Description:\n",
    "# Generates a swarm plot for both case and control data\n",
    "#\n",
    "# Inputs:\n",
    "# input_df: Input data that must have at least DISEASE_COL, ORG_COL\n",
    "# input_case_ax: Axis object for case plot\n",
    "# input_con_ax: Axis object for control plot\n",
    "# DISEASE_COL: Column in input_df where disease status can be found, must\n",
    "#              be encoded as 'case' or 'control'.\n",
    "# ORG_COL: Column in input_df where VIRTUS normalized hit rates for a particular\n",
    "#          pathogen can be found\n",
    "# CASE_X_AXIS_LABEL: X-axis label to go under case plot\n",
    "# CONS_X_AXIS_LABEL: X-axis label to go under control plot\n",
    "# CON_COLOR: Color for control swarm plot markers (RGB)\n",
    "# CASE_COLOR: Color for case swarm plot markers (RGB)\n",
    "# SWARM_MARKER_SIZE: How large the swarm plot markers should be.\n",
    "# SWARM_ALPHA: Opacity of the swarm plot markers\n",
    "# Y_BASELINE: Y-axis lower limit for plot\n",
    "# Y_TICK_LABEL_FD: Font size to use for Y-tick labels\n",
    "# X_TICK_LABEL_FD: Font size to use for label indicating if plot above\n",
    "#                  is a case or control plot.\n",
    "# ANNOT_FD: Font size to output Mann-Whitney P-value label in\n",
    "# ANNOT_LW: Linewidth to draw lines showing what was compared with the\n",
    "#           Mann-Whitney test\n",
    "# DEBUG: Boolean flag whether to print debug messages.\n",
    "\n",
    "# Outputs:\n",
    "#   Returns list of axes\n",
    "#     input_case_ax: Axes containing drawn swarmplot for case data\n",
    "#     input_con_ax:  Axes containing drawn swarmplot for control data\n",
    "\n",
    "from matplotlib.lines import Line2D\n",
    "import matplotlib.transforms as transforms\n",
    "from scipy.stats import mannwhitneyu as mwu\n",
    "from scipy.stats import ttest_ind as ttest\n",
    "from matplotlib.patches import Polygon as poly\n",
    "\n",
    "def gen_swarm_axs(input_df, input_case_ax, input_con_ax,\n",
    "                 DISEASE_COL = 'sample_disease', ORG_COL = 'cmv',\n",
    "                 CASE_X_AXIS_LABEL = 'SLE Cases', \n",
    "                 CONS_X_AXIS_LABEL = 'Controls',\n",
    "                 CON_COLOR =  sns.color_palette()[0],\n",
    "                 CASE_COLOR = sns.color_palette()[1],\n",
    "                 SWARM_MARKER_SIZE = 4, SWARM_ALPHA = 0.7, \n",
    "                 Y_BASELINE = -0.0001,\n",
    "                 Y_TICK_LABEL_FD = {'size' : 11},\n",
    "                 X_TICK_LABEL_FD = {'size' : 11},   \n",
    "                 ANNOT_FD = {'size' : 14},\n",
    "                 ANNOT_LW = 1,\n",
    "                 DEBUG = False):\n",
    "    \n",
    "    \n",
    "    # Generate the case and control labels by first splitting up the data\n",
    "    cases = input_df.loc[input_df[DISEASE_COL] == 'case', :]\n",
    "    cons = input_df.loc[input_df[DISEASE_COL] == 'control', :]\n",
    "\n",
    "    # Counting\n",
    "    case_n = len(cases)\n",
    "    con_n  = len(cons)\n",
    "\n",
    "    # Finally generating label\n",
    "    CASE_X_AXIS_LABEL = f\"{CASE_X_AXIS_LABEL}\\nn = {case_n}\"\n",
    "    CONS_X_AXIS_LABEL = f\"{CONS_X_AXIS_LABEL}\\nn = {con_n}\"\n",
    "\n",
    "    # Make the swarmplots!\n",
    "    \n",
    "    #############################\n",
    "    #         Cases plot        #\n",
    "    #############################\n",
    "    sns.swarmplot(x = DISEASE_COL, y = ORG_COL,\n",
    "                  data = cases, \n",
    "                  color = CASE_COLOR,\n",
    "                  dodge = True,\n",
    "                  size = SWARM_MARKER_SIZE,\n",
    "                  linewidth = 1,\n",
    "                  edgecolor = CASE_COLOR,\n",
    "                  alpha = SWARM_ALPHA,\n",
    "                  ax = input_case_ax)\n",
    "\n",
    "    \n",
    "    # Do some manipulation to case plot\n",
    "    # Set the lower y-axis limit to handed in Y_BASELINE\n",
    "    # Turn off x-axis tick labeling\n",
    "    # Turn of x-axis and y-axis titles\n",
    "    # Turn off the top and right spines of plot\n",
    "    # Adjust the y-axis tick label parameters\n",
    "    #############################\n",
    "    #         Cases-axis        #\n",
    "    #############################\n",
    "    input_case_ax.set_ylim(Y_BASELINE, )\n",
    "    input_case_ax.tick_params(axis = \"x\", \n",
    "                        top = False, labeltop = False,\n",
    "                        bottom = False, labelbottom = False)\n",
    "    \n",
    "    input_case_ax.set_xlabel('')\n",
    "    input_case_ax.set_ylabel('')\n",
    "\n",
    "    \n",
    "\n",
    "    input_case_ax.spines[['top', 'right']].set_visible(False)\n",
    "\n",
    "    input_case_ax.tick_params(axis = \"y\", \n",
    "                    left = True, labelleft = True,\n",
    "                    right = False, labelright = False,\n",
    "                    labelsize = Y_TICK_LABEL_FD['size'])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #############################\n",
    "    #         Cons plot        #\n",
    "    #############################\n",
    "    sns.swarmplot(x = DISEASE_COL, y = ORG_COL,\n",
    "                  data = cons, \n",
    "                  color = CON_COLOR,\n",
    "                  size = SWARM_MARKER_SIZE,\n",
    "                  linewidth = 1,\n",
    "                  edgecolor = CON_COLOR,\n",
    "                  alpha = SWARM_ALPHA,\n",
    "                  ax = input_con_ax)\n",
    "\n",
    "\n",
    "\n",
    "    # Do some manipulation controls plot\n",
    "    # Set the lower y-axis limit to handed in Y_BASELINE\n",
    "    # Turn off x-axis tick labeling\n",
    "    # Turn of x-axis and y-axis titles\n",
    "    # Turn off the top and right spines of plot\n",
    "    # Adjust the y-axis tick label parameters\n",
    "    #############################\n",
    "    #         Cons-axis         #\n",
    "    #############################\n",
    "    input_con_ax.set_ylim(Y_BASELINE, )\n",
    "\n",
    "    input_con_ax.tick_params(axis = \"x\", \n",
    "                        top = False, labeltop = False,\n",
    "                        bottom = False, labelbottom = False)\n",
    "\n",
    "        \n",
    "    input_con_ax.set_xlabel('')\n",
    "    input_con_ax.set_ylabel('')\n",
    "\n",
    "    \n",
    "    input_con_ax.spines[['top', 'right', 'left']].set_visible(False)\n",
    "\n",
    "    input_con_ax.tick_params(axis = \"y\", \n",
    "                    left = False, labelleft = False,\n",
    "                    right = False, labelright = False,\n",
    "                    labelsize = Y_TICK_LABEL_FD['size'])\n",
    "        \n",
    "    \n",
    "    # Calculate the Mann-Whitney p-value\n",
    "    mw_p = mwu(x = cases[ORG_COL], y = cons[ORG_COL], alternative = 'two-sided')[1]\n",
    "\n",
    "\n",
    "    # Add the Case or Control labels under the appropriate plots\n",
    "    CASE_CENTER_X = 0.5\n",
    "    CON_CENTER_X  = 0.5\n",
    "\n",
    "    fig.text(CASE_CENTER_X, -0.025, CASE_X_AXIS_LABEL, \n",
    "             ha='center', va='top', fontdict = X_TICK_LABEL_FD,\n",
    "             transform = input_case_ax.transAxes)\n",
    "    fig.text(CON_CENTER_X, -0.025, CONS_X_AXIS_LABEL, \n",
    "             ha='center', va='top', fontdict = X_TICK_LABEL_FD,\n",
    "             transform = input_con_ax.transAxes)\n",
    "\n",
    "\n",
    "    CASE_CENTER_X = fig.transFigure.inverted().transform(\n",
    "                        input_case_ax.transAxes.transform((0.5, 0.5)))[0]\n",
    "    CON_CENTER_X = fig.transFigure.inverted().transform(\n",
    "                        input_con_ax.transAxes.transform((0.5, 0.5)))[0]\n",
    "\n",
    "    # Now figure out where the Mann-Whitney comparison lines should be drawn\n",
    "    MID_PT = CON_CENTER_X - CASE_CENTER_X\n",
    "    MID_PT = CASE_CENTER_X + ((CON_CENTER_X - CASE_CENTER_X) / 2)\n",
    "\n",
    "    c = max(input_df[ORG_COL])\n",
    "    y_min = 0\n",
    "\n",
    "    y_bot = 0.87\n",
    "    nub_len = 0.025\n",
    "    y_top = y_bot + nub_len\n",
    "\n",
    "    nub_x = CASE_CENTER_X + ((CON_CENTER_X - CASE_CENTER_X) / 2)\n",
    "\n",
    "    x_points = [CASE_CENTER_X, CASE_CENTER_X, MID_PT, MID_PT, MID_PT, CON_CENTER_X, CON_CENTER_X]\n",
    "    y_points = [y_bot, y_top, y_top, y_top + (nub_len/2), y_top, y_top, y_bot]\n",
    "    pts = list(zip(x_points, y_points))\n",
    "\n",
    "    # Draw the line for the Mann-Whitney comparison\n",
    "    line = Line2D(xdata = x_points, ydata = y_points,\n",
    "              linewidth = ANNOT_LW, \n",
    "              linestyle = 'solid', \n",
    "              color = 'black',\n",
    "              clip_on = False,\n",
    "              transform = fig.transFigure)\n",
    "\n",
    "    fig.add_artist(line)\n",
    "\n",
    "\n",
    "    # Add Mann-Whitney p-value text\n",
    "    mw_p_str = \"{:.2g}\".format(mw_p)\n",
    "\n",
    "    fig.text(x = MID_PT, y = 1.02 * (y_top + nub_len/2), \n",
    "             s = f\"Mann-Whitney p = {mw_p_str}\",\n",
    "             fontdict = ANNOT_FD,\n",
    "             ha = 'center')\n",
    "    \n",
    "    \n",
    "    if DEBUG:\n",
    "        print(f'y_axis_title_x_loc: {y_axis_title_x_loc}')\n",
    "    \n",
    "    return [input_case_ax, input_con_ax]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b57f9ca",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.370936Z",
     "start_time": "2023-09-05T18:24:30.345940Z"
    },
    "code_folding": [
     34
    ]
   },
   "outputs": [],
   "source": [
    "# Description:\n",
    "# Generates a bar plot for both diff exp genes and non-diff exp\n",
    "# genes for each GSE ID tested\n",
    "#\n",
    "# Inputs:\n",
    "# input_df: Input data that must have at least DISEASE_NAME column,\n",
    "#           GSE_ID column, gene_type column (DEG/unchanged),\n",
    "#           clean_cell column, neg_log_corr_p column, Enrichment\n",
    "#           column, and P_val column.\n",
    "# input_ax: Axis to draw RELI result bar plots to\n",
    "# DISEASE_NAME: Name of disease column in input_df ['sle', or 'uc']\n",
    "# X_AXIS_LABEL: Label to go under all plots shown in input_ax (such as disease name) \n",
    "# Y_MAX: Max Y value to help in setting the y-ticks\n",
    "# NON_FC: Color for unchanged bars \n",
    "# DIFF_FC: Color for differentially expressed bars\n",
    "# Y_TICK_LABEL_FD: Fontsize for Y tick labels\n",
    "# X_AXIS_TITLE_FD: Fontsize for X_AXIS_LABEL text\n",
    "# X_TICK_LABEL_FD: Fontsize for GSE ID and Cell type labels on X-axis\n",
    "# LEGEND_FONT_SIZE: Fontsize for text in legend\n",
    "# GSE_LAB_Y: Y value to place the GSE ID and Cell type label at\n",
    "# DIS_LAB_Y: Y value to place the X_AXIS_LABEL at\n",
    "# LEG_Y: Y value to place the legend\n",
    "# LEG_X: X value to place the legend\n",
    "# SIG_LW: Linewidth for the signifcance threshold line to draw on plot\n",
    "# ADD_LEGEND: Boolean flag whether to add a legend to the plot\n",
    "# DEBUG: Boolean flag whether to print debug messages.\n",
    "\n",
    "# Outputs:\n",
    "#   Returns input_ax that we have drawn on\n",
    "\n",
    "import matplotlib.transforms as transforms\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.ticker import FormatStrFormatter\n",
    "\n",
    "def gen_reli_axs(input_df, input_ax,\n",
    "                 DISEASE_NAME = 'sle',\n",
    "                 X_AXIS_LABEL = 'SLE Genetic Risk Loci',\n",
    "                 Y_MAX = 10,\n",
    "                 NON_FC =  sns.color_palette()[0],\n",
    "                 DIFF_FC = sns.color_palette()[1],\n",
    "                 Y_TICK_LABEL_FD = {'size' : 11},\n",
    "                 X_AXIS_TITLE_FD = {'size' : 14},\n",
    "                 X_TICK_LABEL_FD = {'size' : 11},   \n",
    "                 LEGEND_FONT_SIZE = 14,\n",
    "                 GSE_LAB_Y = -0.05,\n",
    "                 DIS_LAB_Y = -0.15,\n",
    "                 LEG_Y = 1,\n",
    "                 LEG_X = 1, \n",
    "                 SIG_LW = 1,\n",
    "                 ADD_LEGEND = False,\n",
    "                 DEBUG = False):\n",
    "\n",
    "\n",
    "    \n",
    "    # Options for bars in bar plot\n",
    "    DIFF_EC = 'black'\n",
    "    NON_EC = 'black'\n",
    "    BAR_WIDTH = 1\n",
    "    BAR_SEP = 0.5\n",
    "    GSE_SEP = 0.5\n",
    "    DIS_SEP = 2\n",
    "    \n",
    "    # Options for FC labeling\n",
    "    FOLD_LINE_COLOR = 'black'\n",
    "    LAB_SPACER = 0.5\n",
    "    NUB_DIV = 4\n",
    "\n",
    "\n",
    "    # Grab the data for the disease of interest \n",
    "    filt_dat = input_df.loc[input_df['disease'] == DISEASE_NAME, :]\n",
    "    \n",
    "    # Get list of GSE IDs\n",
    "    filt_gse = input_df['GSE_ID'].unique().tolist()\n",
    "\n",
    "\n",
    "    lab_ls = []\n",
    "    cell_lab_ls = []\n",
    "    fold_ls = []\n",
    "\n",
    "    # Loop through each GSE ID collecting info about it and RELI results\n",
    "    # and creating bar plot\n",
    "    curr_x = 0\n",
    "    for ind, curr_gse in enumerate(filt_gse):\n",
    "        curr_gse_dat = filt_dat.loc[filt_dat['GSE_ID'] == curr_gse, :]\n",
    "\n",
    "\n",
    "        diff_dat = curr_gse_dat.loc[curr_gse_dat['gene_type'] == 'DEG', :]\n",
    "        non_dat = curr_gse_dat.loc[curr_gse_dat['gene_type'] == 'unchanged', :]\n",
    "\n",
    "        curr_cell = diff_dat['clean_cell'].tolist()[0]\n",
    "\n",
    "        diff_neg_p = diff_dat['neg_log_corr_p'].tolist()[0]\n",
    "        diff_enrich = diff_dat['Enrichment'].tolist()[0]\n",
    "        diff_unadj_p = diff_dat['P_val'].tolist()[0]\n",
    "        diff_unadj_neg_log_p = -(np.log10(diff_unadj_p))\n",
    "\n",
    "\n",
    "        non_neg_p = non_dat['neg_log_corr_p'].tolist()[0]\n",
    "        non_enrich = non_dat['Enrichment'].tolist()[0]\n",
    "        non_unadj_p = non_dat['P_val'].tolist()[0]\n",
    "        non_unadj_neg_log_p = -(np.log10(non_unadj_p))\n",
    "\n",
    "        if DEBUG:\n",
    "            print(f\"[gse: {curr_gse}] x: {curr_x}, h: {diff_neg_p}\")\n",
    "\n",
    "        # Diff Bar\n",
    "        input_ax.bar(\n",
    "               x      = curr_x,\n",
    "               height = diff_neg_p,\n",
    "               width  = BAR_WIDTH,\n",
    "               align  = 'center',\n",
    "               fc     = DIFF_FC,\n",
    "               ec     = DIFF_EC\n",
    "               )\n",
    "\n",
    "        mid_gse_x = curr_x + (BAR_WIDTH / 2) + (BAR_SEP / 2)\n",
    "        curr_x = curr_x + BAR_WIDTH + BAR_SEP\n",
    "\n",
    "        lab_ls.append([curr_gse, curr_cell, mid_gse_x])\n",
    "        \n",
    "        if DEBUG:\n",
    "            print(f\"[gse: {curr_gse}] x: {curr_x}, h: {non_neg_p}\")\n",
    "\n",
    "        # UnDiff Bar\n",
    "        input_ax.bar(\n",
    "               x      = curr_x,\n",
    "               height = non_neg_p,\n",
    "               width  = BAR_WIDTH,\n",
    "               align  = 'center',\n",
    "               fc     = NON_FC,\n",
    "               ec     = NON_EC\n",
    "           )\n",
    "        \n",
    "        # Increase the x value taking into account the width of the bar and separator size\n",
    "        curr_x = curr_x + BAR_WIDTH + GSE_SEP\n",
    "        \n",
    "        if DEBUG:\n",
    "            print(f\"[gse: {curr_gse}] x: {curr_x}\")\n",
    "            print(f\"[gse: {curr_gse}]GSE_lab_x: {mid_gse_x}\")\n",
    "            print(\"===========================================================\")\n",
    "\n",
    "   \n",
    "\n",
    "    # Remove x-axis label and tick labels\n",
    "    input_ax.set_xlabel('')\n",
    "    input_ax.tick_params(axis = \"x\", \n",
    "                        top = False, labeltop = False,\n",
    "                        bottom = False, labelbottom = False)\n",
    "\n",
    "    # Remove y-axis label and set ticks based on the Y_MAX value\n",
    "    # Also, remove the top and right spines for plot\n",
    "    input_ax.set_ylabel('')\n",
    "    input_ax.set_yticks(ticks = range(0, Y_MAX + 1, 2))\n",
    "    input_ax.tick_params(axis = \"y\", \n",
    "                            left = True, labelleft = True,\n",
    "                            right = False, labelright = False,\n",
    "                            labelsize = Y_TICK_LABEL_FD['size'])\n",
    "    input_ax.spines[['top', 'right']].set_visible(False)\n",
    "        \n",
    "    \n",
    "    # Label the GSE and cell type\n",
    "    trans = transforms.blended_transform_factory(input_ax.transData, input_ax.transAxes)\n",
    "    for curr_gse, curr_cell, x_val in lab_ls:\n",
    "        input_ax.text(x = x_val,\n",
    "                      y = GSE_LAB_Y,\n",
    "                      s = f\"{curr_gse}\\n[{curr_cell}]\",\n",
    "                      ha = 'center', \n",
    "                      va = 'bottom',\n",
    "                      fontdict = X_TICK_LABEL_FD,\n",
    "                      transform = trans)\n",
    "\n",
    "\n",
    "\n",
    "    # Label the Dis genetic loci\n",
    "    input_ax.text(x  = 0.5, \n",
    "                y  = DIS_LAB_Y,\n",
    "                s  = X_AXIS_LABEL,\n",
    "                ha = 'center',\n",
    "                va = 'bottom',\n",
    "                fontdict = X_AXIS_TITLE_FD,\n",
    "                transform = input_ax.transAxes)\n",
    "\n",
    "\n",
    "    # Add legend\n",
    "    if ADD_LEGEND:\n",
    "\n",
    "        legend_items = [Patch(facecolor = DIFF_FC, edgecolor = DIFF_EC, label = 'Diff Exp Genes'),\n",
    "                        Patch(facecolor = NON_FC, edgecolor = NON_EC, label = 'Unchanged Genes')]\n",
    "\n",
    "\n",
    "        input_ax.legend(handles = legend_items, fontsize = LEGEND_FONT_SIZE,\n",
    "                        loc = 'center', bbox_to_anchor = (LEG_X, LEG_Y),\n",
    "                        ncol = 1, fancybox = True, shadow = False)\n",
    "\n",
    "\n",
    "    # Add 0.05 threshold\n",
    "    thresh = -(np.log10(0.05))\n",
    "    input_ax.axhline(y = thresh, xmin = input_ax.get_xlim()[0], xmax = input_ax.get_xlim()[1],\n",
    "                     linewidth = SIG_LW, linestyle = '--',  color = 'black')\n",
    "\n",
    "    \n",
    "    input_ax.yaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "\n",
    "    return input_ax"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd8cb93",
   "metadata": {},
   "source": [
    "## Get VIRTUS data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7368471a",
   "metadata": {},
   "source": [
    "### SLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0ab1816",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.438071Z",
     "start_time": "2023-09-05T18:24:30.372937Z"
    }
   },
   "outputs": [],
   "source": [
    "# File containing VIRTUS results for all SLE blood and B-cell samples\n",
    "sle_virt_dat = '../virtus/results/sle/all_blood_cleaned_data_sle.csv'\n",
    "\n",
    "dat = pd.read_csv(sle_virt_dat, sep = ',')\n",
    "dat = dat.drop('Unnamed: 0', axis = 1)\n",
    "\n",
    "tdat = dat.transpose()\n",
    "tdat.columns = tdat.loc['org_name', :]\n",
    "\n",
    "# Grab EBV type 1 results\n",
    "ebv = tdat.loc[:, ['ebv_type1', 'sample_disease']]\n",
    "ebv = ebv.drop(['org_name', 'org_grp'], axis = 0)\n",
    "\n",
    "ebv.loc[:, 'ebv_type1'] = ebv.loc[:, 'ebv_type1'].astype(float)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b37fb800",
   "metadata": {},
   "source": [
    "### UC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6961369",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:30.524044Z",
     "start_time": "2023-09-05T18:24:30.440042Z"
    }
   },
   "outputs": [],
   "source": [
    "# File containing VIRTUS results for all UC samples\n",
    "uc_virt_dat = '../virtus/results/cleaned_uc_only_data_no_cov_filter.tsv'\n",
    "\n",
    "uc = pd.read_csv(uc_virt_dat, sep = '\\t')\n",
    "\n",
    "tuc = uc.transpose()\n",
    "tuc.columns = tuc.loc['org_name', :]\n",
    "\n",
    "cmv = tuc.loc[:, ['cmv', 'sample_disease', 'is_ped', 'sample_gse']]\n",
    "cmv = cmv.drop(['org_name', 'org_grp'], axis = 0)\n",
    "\n",
    "cmv.loc[:, 'cmv'] = cmv.loc[:, 'cmv'].astype(float)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19fd737d",
   "metadata": {},
   "source": [
    "## Grab RELI Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9dd3c1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:31.120309Z",
     "start_time": "2023-09-05T18:24:30.526082Z"
    }
   },
   "outputs": [],
   "source": [
    "reli_dat_dir = \"../manuscript/reli_stuff\"\n",
    "\n",
    "# Read in both the SLE and UC RELI results\n",
    "sle_reli  = pd.read_excel(f\"{reli_dat_dir}/2023-06-10_RELI_UC_and_SLE_cleaned_results.xlsx\", sheet_name = 'EBV_SLE')\n",
    "uc_reli = pd.read_excel(f\"{reli_dat_dir}/2023-06-10_RELI_UC_and_SLE_cleaned_results.xlsx\", sheet_name = 'CMV_UC')\n",
    "\n",
    "# Matt: a reasonable cutoff would be corrected P <0.01 and 2-fold change or more.\n",
    "# Filter our results based on Matt's reccomendation\n",
    "sig_sle_reli_gses = sle_reli.loc[((sle_reli['Corrected_P_val'] < 0.01) & \n",
    "                                  (sle_reli['Enrichment'] >= 2)), 'GSE_ID'].tolist()\n",
    "sig_reli_sle = sle_reli.loc[sle_reli['GSE_ID'].isin(sig_sle_reli_gses), :].copy(deep = True)\n",
    "\n",
    "sig_uc_reli_gses = uc_reli.loc[((uc_reli['Corrected_P_val'] < 0.01) & \n",
    "                                (uc_reli['Enrichment'] >= 2)), 'GSE_ID'].tolist()\n",
    "sig_reli_uc = uc_reli.loc[uc_reli['GSE_ID'].isin(sig_uc_reli_gses), :].copy(deep = True)\n",
    "\n",
    "sig_reli_sle.loc[:, 'disease'] = 'sle'\n",
    "sig_reli_uc.loc[:, 'disease'] = 'uc'\n",
    "\n",
    "# Push all RELI results into a single dataframe\n",
    "reli_fin = pd.concat([sig_reli_sle, sig_reli_uc])\n",
    "\n",
    "\n",
    "\n",
    "reli_fin['neg_log_corr_p'] = -(np.log10(reli_fin['Corrected_P_val']))\n",
    "reli_fin = reli_fin.sort_values(['disease', 'neg_log_corr_p'], ascending = [True, False])\n",
    "gse_order = reli_fin['GSE_ID'].drop_duplicates().tolist()\n",
    "\n",
    "# Clean up the cell types\n",
    "reli_fin['clean_cell'] = reli_fin.loc[:, 'Cell_type'].replace({\n",
    "                                                        'monocyte' : 'Monocyte',\n",
    "                                                        'dendritic_cell' : 'Dendritic',\n",
    "                                                        'peripheral_blood_mononuclear_cell' : 'PBMC',\n",
    "                                                        'B-lymphocyte' : 'B-lymphocyte'\n",
    "})\n",
    "\n",
    "# Split them back out by disease\n",
    "fin_reli_sle = reli_fin.loc[reli_fin['disease'] == 'sle', :].copy(deep = True)\n",
    "fin_reli_uc  = reli_fin.loc[reli_fin['disease'] == 'uc', :].copy(deep = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c15edf3",
   "metadata": {},
   "source": [
    "## Start plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d0d354",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:55.626065Z",
     "start_time": "2023-09-05T18:24:31.122277Z"
    },
    "code_folding": [
     0
    ],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# plotting\n",
    "CURR_DEBUG = False\n",
    "\n",
    "# Inches\n",
    "FIG_W = 11\n",
    "FIG_H = 8.5\n",
    "\n",
    "fig = plt.figure(figsize = (FIG_W, FIG_H), facecolor = 'w')\n",
    "\n",
    "# Grid wspace and hspace values\n",
    "WSPACE = 1\n",
    "HSPACE = 1\n",
    "\n",
    "# Serparation between swarm plots in grid slots\n",
    "STATUS_SEP_W = 1\n",
    "\n",
    "# Separator between disease plots in grid slots\n",
    "TEST_SEP_W = 6\n",
    "\n",
    "# Width of swarm plot in grid slots\n",
    "SWARM_WIDTH = 10\n",
    "\n",
    "# Height of swarm plot\n",
    "SWARM_HEIGHT = 10\n",
    "\n",
    "# Separator space between VIRTUS swarm plots and RELI bar plots\n",
    "H_SEP = 2\n",
    "\n",
    "# Split figure grid into 48 columns and 22 rows\n",
    "FIG_GRID_H = 22\n",
    "FIG_GRID_W = SWARM_WIDTH + STATUS_SEP_W + SWARM_WIDTH + TEST_SEP_W + \\\n",
    "                SWARM_WIDTH + STATUS_SEP_W + SWARM_WIDTH\n",
    "\n",
    "\n",
    "# Generate the figure grid\n",
    "fig_grid = plt.GridSpec(FIG_GRID_H, FIG_GRID_W, figure = fig,\n",
    "                        wspace = WSPACE, hspace = HSPACE)\n",
    "\n",
    "############################################\n",
    "#                                          #\n",
    "#         Some parameters                  #\n",
    "#                                          #\n",
    "############################################\n",
    "RELI_DIFF_COLOR = sns.color_palette()[4]\n",
    "RELI_UNDIFF_COLOR   = sns.color_palette()[7]\n",
    "\n",
    "VIRTUS_CASE_COLOR = '#bb3d2e'\n",
    "VIRTUS_CON_COLOR  = 'black'\n",
    "\n",
    "Y_AXIS_TITLE_FD = {'size' : 12}\n",
    "Y_TICK_LABEL_FD = {'size' : 11}\n",
    "X_AXIS_TITLE_FD = {'size' : 12}\n",
    "X_TICK_LABEL_FD = {'size' : 11}\n",
    "ANNOT_FD = {'size' : 10}\n",
    "ANNOT_LW = 1\n",
    "LEGEND_FONT_SIZE = 10\n",
    "\n",
    "\n",
    "LEFT_Y_AXIS_TITLE_OFFSET = 0.0625\n",
    "RIGHT_Y_AXIS_TITLE_OFFSET = 0.0025\n",
    "\n",
    "GSE_LAB_Y = -0.125\n",
    "DIS_LAB_Y = -0.275\n",
    "\n",
    "LEG_X = 0.7\n",
    "LEG_Y = 0.85\n",
    "\n",
    "\n",
    "EBV_VIRT_TEXT = 'EBV Gene Expression Level'\n",
    "CMV_VIRT_TEXT = 'CMV Gene Expression Level'\n",
    "RELI_Y_TEXT   = '-Log10(Corrected RELI p-value)'\n",
    "\n",
    "EBV_RELI_TEXT = 'SLE Genetic Risk Loci overlap with\\nEBV-induced gene expression changes'\n",
    "CMV_RELI_TEXT = 'UC Genetic Risk Loci overlap with\\nCMV-induced gene expression changes'\n",
    "\n",
    "############################################\n",
    "#                                          #\n",
    "#         VIRTUS plots                     #\n",
    "#                                          #\n",
    "############################################\n",
    "\n",
    "# Calculate where each disease and disease status swarm plots should go\n",
    "# in terms of the grid\n",
    "sle_case_swarm_st = 0\n",
    "sle_case_swarm_end = sle_case_swarm_st + SWARM_WIDTH\n",
    "\n",
    "sle_con_swarm_st = sle_case_swarm_end + STATUS_SEP_W\n",
    "sle_con_swarm_end = sle_con_swarm_st + SWARM_WIDTH\n",
    "\n",
    "tot_pl_width = sle_con_swarm_end\n",
    "\n",
    "\n",
    "sle_left_lab_x = (sle_case_swarm_end - sle_case_swarm_st) / 2\n",
    "sle_right_lab_x = (sle_con_swarm_end - sle_con_swarm_st)  / 2\n",
    "\n",
    "uc_case_swarm_st = sle_con_swarm_end + TEST_SEP_W\n",
    "uc_case_swarm_end = uc_case_swarm_st + SWARM_WIDTH\n",
    "\n",
    "uc_con_swarm_st = uc_case_swarm_end + STATUS_SEP_W\n",
    "uc_con_swarm_end = uc_con_swarm_st + SWARM_WIDTH\n",
    "\n",
    "\n",
    "uc_left_lab_x = (uc_case_swarm_end - uc_case_swarm_st) / 2\n",
    "uc_right_lab_x = (uc_con_swarm_end - uc_con_swarm_st)  / 2\n",
    "\n",
    "\n",
    "sle_case_swarm_ax = fig.add_subplot(fig_grid[:SWARM_HEIGHT, sle_case_swarm_st: sle_case_swarm_end])\n",
    "sle_con_swarm_ax = fig.add_subplot(fig_grid[:SWARM_HEIGHT, sle_con_swarm_st : sle_con_swarm_end], sharey = sle_case_swarm_ax)\n",
    "\n",
    "uc_case_swarm_ax = fig.add_subplot(fig_grid[:SWARM_HEIGHT, uc_case_swarm_st: uc_case_swarm_end])\n",
    "uc_con_swarm_ax = fig.add_subplot(fig_grid[:SWARM_HEIGHT, uc_con_swarm_st : uc_con_swarm_end], sharey = uc_case_swarm_ax)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# SLE VIRTUS plot\n",
    "sle_case_swarm_ax, sle_con_swarm_ax = gen_swarm_axs(input_df = ebv,\n",
    "                                                    input_case_ax = sle_case_swarm_ax,\n",
    "                                                    input_con_ax = sle_con_swarm_ax,\n",
    "                                                    DISEASE_COL = 'sample_disease', \n",
    "                                                    ORG_COL = 'ebv_type1',\n",
    "                                                    CASE_X_AXIS_LABEL = 'SLE Cases', \n",
    "                                                    CONS_X_AXIS_LABEL = 'Controls',\n",
    "                                                    CON_COLOR =  VIRTUS_CON_COLOR,\n",
    "                                                    CASE_COLOR = VIRTUS_CASE_COLOR,\n",
    "                                                    SWARM_MARKER_SIZE = 4, \n",
    "                                                    SWARM_ALPHA = 0.7,\n",
    "                                                    Y_BASELINE = -0.0001,\n",
    "                                                    Y_TICK_LABEL_FD = Y_TICK_LABEL_FD,\n",
    "                                                    X_TICK_LABEL_FD = X_TICK_LABEL_FD,\n",
    "                                                    ANNOT_FD = ANNOT_FD, \n",
    "                                                    ANNOT_LW = ANNOT_LW,\n",
    "                                                    DEBUG = CURR_DEBUG)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# UC Adults and Peds VIRTUS plot\n",
    "uc_case_swarm_ax, uc_con_swarm_ax = gen_swarm_axs(input_df = cmv,\n",
    "                                                  input_case_ax = uc_case_swarm_ax,\n",
    "                                                  input_con_ax = uc_con_swarm_ax,\n",
    "                                                  DISEASE_COL = 'sample_disease', \n",
    "                                                  ORG_COL = 'cmv',\n",
    "                                                  CASE_X_AXIS_LABEL = 'UC Cases', \n",
    "                                                  CONS_X_AXIS_LABEL = 'Controls',\n",
    "                                                  CON_COLOR =  VIRTUS_CON_COLOR,\n",
    "                                                  CASE_COLOR = VIRTUS_CASE_COLOR,\n",
    "                                                  SWARM_MARKER_SIZE = 4, \n",
    "                                                  SWARM_ALPHA = 0.7,\n",
    "                                                  Y_BASELINE = -0.001,\n",
    "                                                  Y_TICK_LABEL_FD = Y_TICK_LABEL_FD,\n",
    "                                                  X_TICK_LABEL_FD = X_TICK_LABEL_FD,\n",
    "                                                  ANNOT_FD = ANNOT_FD, \n",
    "                                                  ANNOT_LW = ANNOT_LW, \n",
    "                                                  DEBUG = CURR_DEBUG)\n",
    "    \n",
    "\n",
    "    \n",
    "############################################\n",
    "#                                          #\n",
    "#         RELI plots                       #\n",
    "#                                          #\n",
    "############################################ \n",
    "CURR_DEBUG = False\n",
    "\n",
    "# Calculate where each disease and gene expression status RELI bar plots \n",
    "# should go in terms of the grid\n",
    "PLOT_SEP_W = TEST_SEP_W\n",
    "STATUS_SEP_W = 2\n",
    "\n",
    "\n",
    "sle_reli_st = 0\n",
    "sle_reli_end = sle_reli_st + tot_pl_width\n",
    "\n",
    "uc_reli_st = sle_reli_end + PLOT_SEP_W\n",
    "uc_reli_end = uc_reli_st + tot_pl_width\n",
    "\n",
    "\n",
    "sle_lab_x = (sle_reli_end - sle_reli_st) / 2\n",
    "uc_lab_x = (uc_reli_end - uc_reli_st)  / 2\n",
    "\n",
    "\n",
    "sle_bar_ax = fig.add_subplot(fig_grid[(SWARM_HEIGHT + H_SEP) :, sle_reli_st: sle_reli_end])\n",
    "uc_bar_ax = fig.add_subplot(fig_grid[(SWARM_HEIGHT + H_SEP):, uc_reli_st: uc_reli_end])    \n",
    "    \n",
    "    \n",
    "if CURR_DEBUG:\n",
    "\n",
    "    print(f\"[SLE case swarm]: Height = {SWARM_HEIGHT}, grid_st = {sle_case_swarm_st}, grid_end = {sle_case_swarm_end}\")\n",
    "    print(f\"[SLE con swarm]: Height = {SWARM_HEIGHT}, grid_st = {sle_con_swarm_st}, grid_end = {sle_con_swarm_end}\")\n",
    "    print(f\"[UC case swarm]: Height = {SWARM_HEIGHT}, grid_st = {uc_case_swarm_st}, grid_end = {uc_case_swarm_end}\")\n",
    "    print(f\"[UC con swarm]: Height = {SWARM_HEIGHT}, grid_st = {uc_con_swarm_st}, grid_end = {uc_con_swarm_end}\")\n",
    "    print(f\"[SLE bar]: Height = {SWARM_HEIGHT}, grid_st = {sle_reli_st}, grid_end = {sle_reli_end}\")\n",
    "    print(f\"[UC bar]: Height = {SWARM_HEIGHT}, grid_st = {uc_reli_st}, grid_end = {uc_reli_end}\")\n",
    "\n",
    "    print(FIG_GRID_W)\n",
    "    \n",
    "    \n",
    "sle_bar_ax = gen_reli_axs(input_df = fin_reli_sle, \n",
    "                             input_ax = sle_bar_ax,\n",
    "                             DISEASE_NAME = 'sle',\n",
    "                             X_AXIS_LABEL = EBV_RELI_TEXT,\n",
    "                             Y_MAX = 6,\n",
    "                             NON_FC =  RELI_UNDIFF_COLOR,\n",
    "                             DIFF_FC = RELI_DIFF_COLOR,\n",
    "                             ADD_LEGEND = True,\n",
    "                             DEBUG = CURR_DEBUG,\n",
    "                             Y_TICK_LABEL_FD = Y_TICK_LABEL_FD,\n",
    "                             X_AXIS_TITLE_FD = X_AXIS_TITLE_FD,\n",
    "                             X_TICK_LABEL_FD = X_TICK_LABEL_FD,\n",
    "                             LEGEND_FONT_SIZE = LEGEND_FONT_SIZE,\n",
    "                             GSE_LAB_Y = GSE_LAB_Y,\n",
    "                             DIS_LAB_Y = DIS_LAB_Y,\n",
    "                             SIG_LW = 2,\n",
    "                             LEG_Y = LEG_Y, \n",
    "                             LEG_X = LEG_X)\n",
    "\n",
    "\n",
    "uc_bar_ax = gen_reli_axs(input_df = fin_reli_uc, \n",
    "                             input_ax = uc_bar_ax,\n",
    "                             DISEASE_NAME = 'uc',\n",
    "                             X_AXIS_LABEL = CMV_RELI_TEXT,\n",
    "                             Y_MAX = 10,\n",
    "                             NON_FC =  RELI_UNDIFF_COLOR,\n",
    "                             DIFF_FC = RELI_DIFF_COLOR,\n",
    "                             ADD_LEGEND = False,\n",
    "                             DEBUG = CURR_DEBUG,\n",
    "                             Y_TICK_LABEL_FD = Y_TICK_LABEL_FD,\n",
    "                             X_AXIS_TITLE_FD = X_AXIS_TITLE_FD,\n",
    "                             X_TICK_LABEL_FD = X_TICK_LABEL_FD,\n",
    "                             LEGEND_FONT_SIZE = LEGEND_FONT_SIZE,\n",
    "                             GSE_LAB_Y = GSE_LAB_Y,\n",
    "                             DIS_LAB_Y = DIS_LAB_Y,\n",
    "                             SIG_LW = 2,\n",
    "                             LEG_Y = LEG_Y, \n",
    "                             LEG_X = LEG_X)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig.text(x = 0 + LEFT_Y_AXIS_TITLE_OFFSET, \n",
    "         y = fig.transFigure.inverted().transform(sle_case_swarm_ax.transAxes.transform((0, 0.5)))[1],\n",
    "                s = EBV_VIRT_TEXT, \n",
    "                fontdict = Y_AXIS_TITLE_FD,\n",
    "                ha = 'center', va = 'center', \n",
    "                rotation = 90)\n",
    "\n",
    "\n",
    "fig.text(x = 0.5 - RIGHT_Y_AXIS_TITLE_OFFSET, \n",
    "         y = fig.transFigure.inverted().transform(uc_con_swarm_ax.transAxes.transform((0, 0.5)))[1],\n",
    "                s = CMV_VIRT_TEXT, \n",
    "                fontdict = Y_AXIS_TITLE_FD,\n",
    "                ha = 'center', va = 'center', \n",
    "                rotation = 90)\n",
    "\n",
    "\n",
    "fig.text(x = 0 + LEFT_Y_AXIS_TITLE_OFFSET, \n",
    "         y = fig.transFigure.inverted().transform(sle_bar_ax.transAxes.transform((0, 0.5)))[1],\n",
    "                s = RELI_Y_TEXT, \n",
    "                fontdict = Y_AXIS_TITLE_FD,\n",
    "                ha = 'center', va = 'center', \n",
    "                rotation = 90)\n",
    "\n",
    "fig.text(x = 0.5 - RIGHT_Y_AXIS_TITLE_OFFSET, \n",
    "         y = fig.transFigure.inverted().transform(uc_bar_ax.transAxes.transform((0, 0.5)))[1],\n",
    "                s = RELI_Y_TEXT, \n",
    "                fontdict = Y_AXIS_TITLE_FD,\n",
    "                ha = 'center', va = 'center', \n",
    "                rotation = 90)\n",
    "\n",
    "\n",
    "HALF_Y = 0.475\n",
    "TOP_Y  = 0.95\n",
    "\n",
    "LEFT_X = 0.025\n",
    "HALF_X = 0.5\n",
    "\n",
    "\n",
    "\n",
    "# Add panel letter labels\n",
    "panel_font = {'size' : 20, 'weight' : 'bold'}\n",
    "\n",
    "fig.text(x = LEFT_X, y = TOP_Y, s = 'A', fontdict = panel_font)\n",
    "fig.text(x = HALF_X, y = TOP_Y, s = 'B', fontdict = panel_font)\n",
    "fig.text(x = LEFT_X, y = HALF_Y, s = 'C', fontdict = panel_font)\n",
    "fig.text(x = HALF_X, y = HALF_Y, s = 'D', fontdict = panel_font)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4f9bec2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-09-05T18:24:56.501196Z",
     "start_time": "2023-09-05T18:24:55.628037Z"
    }
   },
   "outputs": [],
   "source": [
    "out_dir = '../manuscript/figures/fig_5'\n",
    "fn = f\"{out_dir}/fig_5_pub.pdf\"\n",
    "fig.savefig(fn, format = 'pdf', dpi = 600, bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
